<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<title>RNAlib-2.0.7: Folding Routines - Functions for Folding RNA Secondary Structures</title>
<link href="tabs.css" rel="stylesheet" type="text/css"/>
<link href="navtree.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="jquery.js"></script>
<script type="text/javascript" src="navtree.js"></script>
<script type="text/javascript" src="resize.js"></script>
<script type="text/javascript">
$(document).ready(initResizable);
</script>
<link href="doxygen.css" rel="stylesheet" type="text/css"/>
</head>
<body>
<!-- Generated by Doxygen 1.7.4 -->
<div id="top">
<div id="titlearea">
<table cellspacing="0" cellpadding="0">
 <tbody>
 <tr style="height: 56px;">
  <td style="padding-left: 0.5em;">
   <div id="projectname">RNAlib-2.0.7</div>
  </td>
 </tr>
 </tbody>
</table>
</div>
  <div id="navrow1" class="tabs">
    <ul class="tablist">
      <li><a href="index.html"><span>Main&#160;Page</span></a></li>
      <li class="current"><a href="pages.html"><span>Related&#160;Pages</span></a></li>
      <li><a href="annotated.html"><span>Data&#160;Structures</span></a></li>
      <li><a href="files.html"><span>Files</span></a></li>
    </ul>
  </div>
</div>
<div id="side-nav" class="ui-resizable side-nav-resizable">
  <div id="nav-tree">
    <div id="nav-tree-contents">
    </div>
  </div>
  <div id="splitbar" style="-moz-user-select:none;" 
       class="ui-resizable-handle">
  </div>
</div>
<script type="text/javascript">
  initNavTree('mp_fold.html','');
</script>
<div id="doc-content">
<div class="header">
  <div class="headertitle">
<div class="title">Folding Routines - Functions for Folding RNA Secondary Structures </div>  </div>
</div>
<div class="contents">
<div class="textblock"><p><a class="anchor" id="toc"></a></p>
<h3>Table of Contents</h3>
<hr/>
<ul>
<li><a class="el" href="mp_fold.html#mp_mfe_Fold">Calculating Minimum Free Energy Structures</a> </li>
<li><a class="el" href="mp_fold.html#mp_PF_Fold">Calculating Partition Functions and Pair Probabilities</a> </li>
<li><a class="el" href="mp_fold.html#mp_Inverse_Fold">Searching for Predefined Structures</a> </li>
<li><a class="el" href="mp_fold.html#mp_Suboptimal_folding">Enumerating Suboptimal Structures</a> </li>
<li><a class="el" href="mp_fold.html#mp_Cofolding">Predicting hybridization structures of two molecules</a> </li>
<li><a class="el" href="mp_fold.html#mp_Local_Fold">Predicting local structures of large sequences</a> </li>
<li><a class="el" href="mp_fold.html#mp_Alignment_Fold">Predicting Consensus Structures from Alignment</a> </li>
<li><a class="el" href="mp_fold.html#mp_Fold_Vars">Global Variables for the Folding Routines</a> </li>
<li><a class="el" href="mp_fold.html#mp_Param_Files">Reading Energy Parameters from File</a></li>
</ul>
<hr/>
<h2><a class="anchor" id="mp_mfe_Fold"></a>
Calculating Minimum Free Energy Structures</h2>
<p>The library provides a fast dynamic programming minimum free energy folding algorithm as described by <a class="el" href="mp_ref.html#zuker_81">Zuker &amp; Stiegler (1981)</a>.</p>
<p>Associated functions are:</p>
<div class="fragment"><pre class="fragment">
float fold (char* sequence, char* structure);
</pre></div> <p>Compute minimum free energy and an appropriate secondary structure of an RNA sequence. </p>
<div class="fragment"><pre class="fragment">
float circfold (char* sequence, char* structure);
</pre></div> <p>Compute minimum free energy and an appropriate secondary structure of a circular RNA sequence. </p>
<div class="fragment"><pre class="fragment">
float energy_of_structure(const char *string,
                          const char *structure,
                          int verbosity_level);
</pre></div> <p>Calculate the free energy of an already folded RNA using global model detail settings. </p>
<div class="fragment"><pre class="fragment">
float energy_of_circ_structure( const char *string,
                                const char *structure,
                                int verbosity_level);
</pre></div> <p>Calculate the free energy of an already folded circular RNA. </p>
<div class="fragment"><pre class="fragment">
void  update_fold_params(void);
</pre></div> <p>Recalculate energy parameters. </p>
<div class="fragment"><pre class="fragment">
void  free_arrays(void);
</pre></div> <p>Free arrays for mfe folding. </p>
<dl class="see"><dt><b>See also:</b></dt><dd><a class="el" href="fold_8h.html" title="MFE calculations and energy evaluations for single RNA sequences.">fold.h</a>, <a class="el" href="cofold_8h.html" title="MFE version of cofolding routines.">cofold.h</a>, <a class="el" href="2Dfold_8h.html" title="Compute the minimum free energy (MFE) and secondary structures for a partitioning of the secondary st...">2Dfold.h</a>, <a class="el" href="Lfold_8h.html" title="Predicting local MFE structures of large sequences.">Lfold.h</a>, <a class="el" href="alifold_8h.html" title="compute various properties (consensus MFE structures, partition function, Boltzmann distributed stoch...">alifold.h</a> and <a class="el" href="subopt_8h.html" title="RNAsubopt and density of states declarations.">subopt.h</a> for a complete list of available functions.</dd></dl>
 
<hr>
<a href="#toc">Table of Contents</a>
<hr>
<h2><a class="anchor" id="mp_PF_Fold"></a>
Calculating Partition Functions and Pair Probabilities</h2>
<p>Instead of the minimum free energy structure the partition function of all possible structures and from that the pairing probability for every possible pair can be calculated, using a dynamic programming algorithm as described by <a class="el" href="mp_ref.html#mccaskill_90">McCaskill (1990)</a>. The following functions are provided:</p>
<div class="fragment"><pre class="fragment">
float pf_fold ( char* sequence,
                char* structure)
</pre></div> <p>Compute the partition function <img class="formulaInl" alt="$Q$" src="form_14.png"/> of an RNA sequence. </p>
<div class="fragment"><pre class="fragment">
void free_pf_arrays (void)
</pre></div> <p>Free arrays for the partition function recursions. </p>
<div class="fragment"><pre class="fragment">
void update_pf_params (int length)
</pre></div> <p>Recalculate energy parameters. </p>
<div class="fragment"><pre class="fragment">
char *get_centroid_struct_pl( int length,
                              double *dist,
                              plist *pl);
</pre></div> <p>Get the centroid structure of the ensemble. </p>
<div class="fragment"><pre class="fragment">
char *get_centroid_struct_pr( int length,
                              double *dist,
                              double *pr);
</pre></div> <p>Get the centroid structure of the ensemble. </p>
<div class="fragment"><pre class="fragment">
double mean_bp_distance_pr( int length,
                            double *pr);
</pre></div> <p>Get the mean base pair distance in the thermodynamic ensemble. </p>
<dl class="see"><dt><b>See also:</b></dt><dd><a class="el" href="part__func_8h.html" title="Partition function of single RNA sequences.">part_func.h</a>, <a class="el" href="part__func__co_8h.html" title="Partition function for two RNA sequences.">part_func_co.h</a>, <a class="el" href="part__func__up_8h.html" title="Partition Function Cofolding as stepwise process.">part_func_up.h</a>, <a class="el" href="2Dpfold_8h.html" title="Compute the partition function and stochastically sample secondary structures for a partitioning of t...">2Dpfold.h</a>, <a class="el" href="LPfold_8h.html" title="Function declarations of partition function variants of the Lfold algorithm.">LPfold.h</a>, <a class="el" href="alifold_8h.html" title="compute various properties (consensus MFE structures, partition function, Boltzmann distributed stoch...">alifold.h</a> and <a class="el" href="MEA_8h.html" title="Computes a MEA (maximum expected accuracy) structure.">MEA.h</a> for a complete list of available functions.</dd></dl>
 
<hr>
<a href="#toc">Table of Contents</a>
<hr>
<h2><a class="anchor" id="mp_Inverse_Fold"></a>
Searching for Predefined Structures</h2>
<p>We provide two functions that search for sequences with a given structure, thereby inverting the folding routines.</p>
<div class="fragment"><pre class="fragment">
float inverse_fold (char *start,
                    char *target)
</pre></div> <p>Find sequences with predefined structure. </p>
<div class="fragment"><pre class="fragment">
float inverse_pf_fold ( char *start,
                        char *target)
</pre></div> <p>Find sequence that maximizes probability of a predefined structure. </p>
<p>The following global variables define the behavior or show the results of the inverse folding routines:</p>
<div class="fragment"><pre class="fragment">
char *symbolset
</pre></div> <p>This global variable points to the allowed bases, initially "AUGC". </p>
<dl class="see"><dt><b>See also:</b></dt><dd><a class="el" href="inverse_8h.html" title="Inverse folding routines.">inverse.h</a> for more details and a complete list of available functions.</dd></dl>
 
<hr>
<a href="#toc">Table of Contents</a>
<hr>
<h2><a class="anchor" id="mp_Suboptimal_folding"></a>
Enumerating Suboptimal Structures</h2>
<div class="fragment"><pre class="fragment">
SOLUTION *subopt (char *sequence,
                  char *constraint,
                  int *delta,
                  FILE *fp)
</pre></div> <p>Returns list of subopt structures or writes to fp. </p>
<div class="fragment"><pre class="fragment">
SOLUTION *subopt_circ ( char *sequence,
                        char *constraint,
                        int *delta,
                        FILE *fp)
</pre></div> <p>Returns list of circular subopt structures or writes to fp. </p>
<div class="fragment"><pre class="fragment">
SOLUTION  *zukersubopt(const char *string);
</pre></div> <p>Compute Zuker type suboptimal structures. </p>
<div class="fragment"><pre class="fragment">
char  *TwoDpfold_pbacktrack ( TwoDpfold_vars *vars,
                              unsigned int d1,
                              unsigned int d2)
</pre></div> <p>Sample secondary structure representatives from a set of distance classes according to their Boltzmann probability. </p>
<div class="fragment"><pre class="fragment">
char  *alipbacktrack (double *prob)
</pre></div> <p>Sample a consensus secondary structure from the Boltzmann ensemble according its probability<br/>
. </p>
<div class="fragment"><pre class="fragment">
char    *pbacktrack(char *sequence);
</pre></div> <p>Sample a secondary structure from the Boltzmann ensemble according its probability<br/>
. </p>
<div class="fragment"><pre class="fragment">
char    *pbacktrack_circ(char *sequence);
</pre></div> <p>Sample a secondary structure of a circular RNA from the Boltzmann ensemble according its probability. </p>
<dl class="see"><dt><b>See also:</b></dt><dd><a class="el" href="subopt_8h.html" title="RNAsubopt and density of states declarations.">subopt.h</a>, <a class="el" href="part__func_8h.html" title="Partition function of single RNA sequences.">part_func.h</a>, <a class="el" href="alifold_8h.html" title="compute various properties (consensus MFE structures, partition function, Boltzmann distributed stoch...">alifold.h</a> and <a class="el" href="2Dpfold_8h.html" title="Compute the partition function and stochastically sample secondary structures for a partitioning of t...">2Dpfold.h</a> for more detailed descriptions</dd></dl>
 
<hr>
<a href="#toc">Table of Contents</a>
<hr>
<h2><a class="anchor" id="mp_Cofolding"></a>
Predicting hybridization structures of two molecules</h2>
<p>The function of an RNA molecule often depends on its interaction with other RNAs. The following routines therefore allow to predict structures formed by two RNA molecules upon hybridization.<br/>
 One approach to co-folding two RNAs consists of concatenating the two sequences and keeping track of the concatenation point in all energy evaluations. Correspondingly, many of the <a class="el" href="cofold_8h.html#abc8517f22cfe70595ee81fc837910d52" title="Compute the minimum free energy of two interacting RNA molecules.">cofold()</a> and <a class="el" href="part__func__co_8h.html#aa86a5f998789ed71813d23d7307a791b" title="Calculate partition function and base pair probabilities.">co_pf_fold()</a> routines below take one sequence string as argument and use the the global variable <a class="el" href="fold__vars_8h.html#ab9b2c3a37a5516614c06d0ab54b97cda" title="Marks the position (starting from 1) of the first nucleotide of the second molecule within the concat...">cut_point</a> to mark the concatenation point. Note that while the <em>RNAcofold</em> program uses the '&amp;' character to mark the chain break in its input, you should not use an '&amp;' when using the library routines (set <a class="el" href="fold__vars_8h.html#ab9b2c3a37a5516614c06d0ab54b97cda" title="Marks the position (starting from 1) of the first nucleotide of the second molecule within the concat...">cut_point</a> instead).<br/>
 In a second approach to co-folding two RNAs, cofolding is seen as a stepwise process. In the first step the probability of an unpaired region is calculated and in a second step this probability of an unpaired region is multiplied with the probability of an interaction between the two RNAs. This approach is implemented for the interaction between a long target sequence and a short ligand RNA. Function <a class="el" href="part__func__up_8h.html#a5b4ee40e190d2f633cd01cf0d2fe93cf" title="Calculate the partition function over all unpaired regions of a maximal length.">pf_unstru()</a> calculates the partition function over all unpaired regions in the input sequence. Function <a class="el" href="part__func__up_8h.html#a1aa0aa02bc3a724f87360c03097afd00" title="Calculates the probability of a local interaction between two sequences.">pf_interact()</a>, which calculates the partition function over all possible interactions between two sequences, needs both sequence as separate strings as input.</p>
<div class="fragment"><pre class="fragment">
int cut_point
</pre></div> <p>Marks the position (starting from 1) of the first nucleotide of the second molecule within the concatenated sequence. </p>
<div class="fragment"><pre class="fragment">
float cofold (char *sequence,
              char *structure)
</pre></div> <p>Compute the minimum free energy of two interacting RNA molecules. </p>
<div class="fragment"><pre class="fragment">
void  free_co_arrays (void)
</pre></div> <p>Free memory occupied by <a class="el" href="cofold_8h.html#abc8517f22cfe70595ee81fc837910d52" title="Compute the minimum free energy of two interacting RNA molecules.">cofold()</a> </p>
<p><b>Partition Function Cofolding</b></p>
<p>To simplify the implementation the partition function computation is done internally in a null model that does not include the duplex initiation energy, i.e. the entropic penalty for producing a dimer from two monomers). The resulting free energies and pair probabilities are initially relative to that null model. In a second step the free energies can be corrected to include the dimerization penalty, and the pair probabilities can be divided into the conditional pair probabilities given that a re dimer is formed or not formed.</p>
<div class="fragment"><pre class="fragment">
cofoldF co_pf_fold( char *sequence,
                    char *structure);
</pre></div> <p>Calculate partition function and base pair probabilities. </p>
<div class="fragment"><pre class="fragment">
void    free_co_pf_arrays(void);
</pre></div> <p>Free the memory occupied by <a class="el" href="part__func__co_8h.html#aa86a5f998789ed71813d23d7307a791b" title="Calculate partition function and base pair probabilities.">co_pf_fold()</a> </p>
<p><b>Cofolding all Dimeres, Concentrations</b></p>
<p>After computing the partition functions of all possible dimeres one can compute the probabilities of base pairs, the concentrations out of start concentrations and sofar and soaway.</p>
<div class="fragment"><pre class="fragment">
void  compute_probabilities(
              double FAB,
              double FEA,
              double FEB,
              struct plist  *prAB,
              struct plist  *prA,
              struct plist  *prB,
              int Alength)
</pre></div> <p>Compute Boltzmann probabilities of dimerization without homodimers. </p>
<div class="fragment"><pre class="fragment">
ConcEnt *get_concentrations(double FEAB,
                            double FEAA,
                            double FEBB,
                            double FEA,
                            double FEB,
                            double * startconc)
</pre></div> <p>Given two start monomer concentrations a and b, compute the concentrations in thermodynamic equilibrium of all dimers and the monomers. </p>
<p><b>Partition Function Cofolding as a stepwise process</b></p>
<p>In this approach to cofolding the interaction between two RNA molecules is seen as a stepwise process. In a first step, the target molecule has to adopt a structure in which a binding site is accessible. In a second step, the ligand molecule will hybridize with a region accessible to an interaction. Consequently the algorithm is designed as a two step process: The first step is the calculation of the probability that a region within the target is unpaired, or equivalently, the calculation of the free energy needed to expose a region. In the second step we compute the free energy of an interaction for every possible binding site. Associated functions are:</p>
<div class="fragment"><pre class="fragment">
pu_contrib *pf_unstru ( char *sequence,
                        int max_w)
</pre></div> <p>Calculate the partition function over all unpaired regions of a maximal length. </p>
<div class="fragment"><pre class="fragment">
void  free_pu_contrib_struct (pu_contrib *pu)
</pre></div> <p>Frees the output of function <a class="el" href="part__func__up_8h.html#a5b4ee40e190d2f633cd01cf0d2fe93cf" title="Calculate the partition function over all unpaired regions of a maximal length.">pf_unstru()</a>. </p>
<div class="fragment"><pre class="fragment">
interact *pf_interact(
              const char *s1,
              const char *s2,
              pu_contrib *p_c,
              pu_contrib *p_c2,
              int max_w,
              char *cstruc,
              int incr3,
              int incr5)
</pre></div> <p>Calculates the probability of a local interaction between two sequences. </p>
<div class="fragment"><pre class="fragment">
void free_interact (interact *pin)
</pre></div> <p>Frees the output of function <a class="el" href="part__func__up_8h.html#a1aa0aa02bc3a724f87360c03097afd00" title="Calculates the probability of a local interaction between two sequences.">pf_interact()</a>. </p>
<dl class="see"><dt><b>See also:</b></dt><dd><a class="el" href="cofold_8h.html" title="MFE version of cofolding routines.">cofold.h</a>, <a class="el" href="part__func__co_8h.html" title="Partition function for two RNA sequences.">part_func_co.h</a> and <a class="el" href="part__func__up_8h.html" title="Partition Function Cofolding as stepwise process.">part_func_up.h</a> for more details</dd></dl>
 
<hr>
<a href="#toc">Table of Contents</a>
<hr>
<h2><a class="anchor" id="mp_Local_Fold"></a>
Predicting local structures of large sequences</h2>
<p>Local structures can be predicted by a modified version of the <a class="el" href="fold_8h.html#aadafcb0f140795ae62e5ca027e335a9b" title="Compute minimum free energy and an appropriate secondary structure of an RNA sequence.">fold()</a> algorithm that restricts the span of all base pairs.</p>
<div class="fragment"><pre class="fragment">
float Lfold ( const char *string,
              char *structure,
              int maxdist)
</pre></div> <p>The local analog to <a class="el" href="fold_8h.html#aadafcb0f140795ae62e5ca027e335a9b" title="Compute minimum free energy and an appropriate secondary structure of an RNA sequence.">fold()</a>. </p>
<div class="fragment"><pre class="fragment">
float aliLfold( const char **strings,
                char *structure,
                int maxdist)
</pre></div> <div class="fragment"><pre class="fragment">
float Lfoldz (const char *string,
              char *structure,
              int maxdist,
              int zsc,
              double min_z)
</pre></div> <div class="fragment"><pre class="fragment">
plist *pfl_fold (
            char *sequence,
            int winSize,
            int pairSize,
            float cutoffb,
            double **pU,
            struct plist **dpp2,
            FILE *pUfp,
            FILE *spup)
</pre></div> <p>Compute partition functions for locally stable secondary structures <b>(berni! update me)</b> </p>
<dl class="see"><dt><b>See also:</b></dt><dd><a class="el" href="Lfold_8h.html" title="Predicting local MFE structures of large sequences.">Lfold.h</a> and <a class="el" href="LPfold_8h.html" title="Function declarations of partition function variants of the Lfold algorithm.">LPfold.h</a> for more details</dd></dl>
 
<hr>
<a href="#toc">Table of Contents</a>
<hr>
<h2><a class="anchor" id="mp_Alignment_Fold"></a>
Predicting Consensus Structures from Alignment</h2>
<p>Consensus structures can be predicted by a modified version of the <a class="el" href="fold_8h.html#aadafcb0f140795ae62e5ca027e335a9b" title="Compute minimum free energy and an appropriate secondary structure of an RNA sequence.">fold()</a> algorithm that takes a set of aligned sequences instead of a single sequence. The energy function consists of the mean energy averaged over the sequences, plus a covariance term that favors pairs with consistent and compensatory mutations and penalizes pairs that cannot be formed by all structures. For details see <a class="el" href="mp_ref.html#hofacker_02">Hofacker (2002)</a>.</p>
<div class="fragment"><pre class="fragment">
float  alifold (const char **strings,
                char *structure)
</pre></div> <p>Compute MFE and according consensus structure of an alignment of sequences. </p>
<div class="fragment"><pre class="fragment">
float  circalifold (const char **strings,
                    char *structure)
</pre></div> <p>Compute MFE and according structure of an alignment of sequences assuming the sequences are circular instead of linear. </p>
<div class="fragment"><pre class="fragment">
void    free_alifold_arrays (void)
</pre></div> <p>Free the memory occupied by MFE alifold functions. </p>
<div class="fragment"><pre class="fragment">
float   energy_of_alistruct (
            const char **sequences,
            const char *structure,
            int n_seq,
            float *energy)
</pre></div> <p>Calculate the free energy of a consensus structure given a set of aligned sequences. </p>
<div class="fragment"><pre class="fragment">
struct pair_info
</pre></div> <p>A base pair info structure. </p>
<div class="fragment"><pre class="fragment">
double  cv_fact
</pre></div> <p>This variable controls the weight of the covariance term in the energy function of alignment folding algorithms. </p>
<div class="fragment"><pre class="fragment">
double  nc_fact
</pre></div> <p>This variable controls the magnitude of the penalty for non-compatible sequences in the covariance term of alignment folding algorithms. </p>
<dl class="see"><dt><b>See also:</b></dt><dd><a class="el" href="alifold_8h.html" title="compute various properties (consensus MFE structures, partition function, Boltzmann distributed stoch...">alifold.h</a> for more details</dd></dl>
 
<hr>
<a href="#toc">Table of Contents</a>
<hr>
<h2><a class="anchor" id="mp_Fold_Vars"></a>
Global Variables for the Folding Routines</h2>
<p>The following global variables change the behavior the folding algorithms or contain additional information after folding.</p>
<div class="fragment"><pre class="fragment">
int noGU
</pre></div> <p>Global switch to forbid/allow GU base pairs at all. </p>
<div class="fragment"><pre class="fragment">
int no_closingGU
</pre></div> <p>GU allowed only inside stacks if set to 1. </p>
<div class="fragment"><pre class="fragment">
int noLonelyPairs
</pre></div> <p>Global switch to avoid/allow helices of length 1. </p>
<div class="fragment"><pre class="fragment">
int tetra_loop
</pre></div> <p>Include special stabilizing energies for some tri-, tetra- and hexa-loops;. </p>
<div class="fragment"><pre class="fragment">
int energy_set
</pre></div> <p>0 = BP; 1=any mit GC; 2=any mit AU-parameter </p>
<div class="fragment"><pre class="fragment">
float temperature
</pre></div> <p>Rescale energy parameters to a temperature in degC. </p>
<div class="fragment"><pre class="fragment">
int dangles
</pre></div> <p>Switch the energy model for dangling end contributions (0, 1, 2, 3) </p>
<div class="fragment"><pre class="fragment">
char *nonstandards
</pre></div> <p>contains allowed non standard base pairs </p>
<div class="fragment"><pre class="fragment">
int cut_point
</pre></div> <p>Marks the position (starting from 1) of the first nucleotide of the second molecule within the concatenated sequence. </p>
<div class="fragment"><pre class="fragment">
float pf_scale
</pre></div> <p>A scaling factor used by <a class="el" href="part__func_8h.html#adc3db3d98742427e7001a7fd36ef28c2" title="Compute the partition function  of an RNA sequence.">pf_fold()</a> to avoid overflows. </p>
<div class="fragment"><pre class="fragment">
int fold_constrained
</pre></div> <p>Global switch to activate/deactivate folding with structure constraints. </p>
<div class="fragment"><pre class="fragment">
int do_backtrack
</pre></div> <p>do backtracking, i.e. </p>
<div class="fragment"><pre class="fragment">
char backtrack_type
</pre></div> <p>A backtrack array marker for <a class="el" href="inverse_8h.html#a7af026de55d4babad879f2c92559cbbc" title="Find sequences with predefined structure.">inverse_fold()</a> </p>
<p>include <a class="el" href="fold__vars_8h.html" title="Here all all declarations of the global variables used throughout RNAlib.">fold_vars.h</a> if you want to change any of these variables from their defaults.</p>
<dl class="see"><dt><b>See also:</b></dt><dd><a class="el" href="fold__vars_8h.html" title="Here all all declarations of the global variables used throughout RNAlib.">fold_vars.h</a> for a more complete and detailed description of all global variables and how to use them</dd></dl>
 
<hr>
<a href="#toc">Table of Contents</a>
<hr>
<h2><a class="anchor" id="mp_Param_Files"></a>
Reading Energy Parameters from File</h2>
<p>A default set of parameters, identical to the one described in <a class="el" href="mp_ref.html#mathews_04">Mathews et.al. (2004)</a>, is compiled into the library.<br/>
 Alternately, parameters can be read from and written to a file.</p>
<div class="fragment"><pre class="fragment">
void  read_parameter_file (const char fname[])
</pre></div> <p>Read energy parameters from a file. </p>
<div class="fragment"><pre class="fragment">
void  write_parameter_file (const char fname[])
</pre></div> <p>Write energy parameters to a file. </p>
<p>To preserve some backward compatibility the RNAlib also provides functions to convert energy parameter files from the format used in version 1.4-1.8 into the new format used since version 2.0</p>
<div class="fragment"><pre class="fragment">
void convert_parameter_file (
            const char *iname,
            const char *oname,
            unsigned int options)
</pre></div> <p>Convert/dump a Vienna 1.8.4 formatted energy parameter file. </p>
<dl class="see"><dt><b>See also:</b></dt><dd><a class="el" href="read__epars_8h.html" title="Functions to read and write energy parameter sets from/to files.">read_epars.h</a> and <a class="el" href="convert__epars_8h.html" title="Functions and definitions for energy parameter file format conversion.">convert_epars.h</a> for detailed description of the available functions</dd></dl>
 
<hr>
<a href="#toc">Table of Contents</a>
<hr>
<p><a class="el" href="mp_parse.html">Next Page: Parsing and Comparing</a> </p>
</div></div>
</div>
  <div id="nav-path" class="navpath">
    <ul>
      <li class="footer">Generated on Fri May 18 2012 for RNAlib-2.0.7 by&#160;
<a href="http://www.doxygen.org/index.html">
<img class="footer" src="doxygen.png" alt="doxygen"/></a> 1.7.4 </li>
    </ul>
  </div>

</body>
</html>
