{% extends "layout.html" %}
{% block content %}
    <script type="text/javascript"
            src="https://cdn.jsdelivr.net/npm/igv@2.10.4/dist/igv.min.js"></script>
    <head>
        <link rel="stylesheet"
              href="{{ url_for('static', filename='css/common.css') }}">
        <link rel="stylesheet"
              href="{{ url_for('static', filename='css/triplet_periodicity.css') }}">
    </head>
    <style>
.tab {
	overflow:hidden;
	border: 1px solid #ccc;
	background-color: #f1f1f1;
	text-align: center;

}

.tab button {
	background-color: inherit;
	border: none;
	outline: none;
	cursor: pointer;
	padding: 14px 16px;
	transition: 0.3s;
}


.tab button:hover {
	background-color: #ddd;
}

.tab button.active {
	background-color: #ccc;
}

.tabcontent {
	display: none;
	padding: 6px 12px;
	border: 1px solid #ccc;
	border-top: none;
}

.wrap
{
    width: 320px;
    height: 192px;
    padding: 0;
    overflow: hidden;
}

.frame
{
    width: 1200px;
    height: 786px;
    border: 0;

    -ms-transform: scale(0.25);
    -moz-transform: scale(0.25);
    -o-transform: scale(0.25);
    -webkit-transform: scale(0.25);
    transform: scale(0.25);

    -ms-transform-origin: 0 0;
    -moz-transform-origin: 0 0;
    -o-transform-origin: 0 0;
    -webkit-transform-origin: 0 0;
    transform-origin: 0 0;
}

    </style>
    <div id="page_heading">
        <p>
            <font size="5"><b><u>General Settings</u></b></font>
        </p>
    </div>
    <div id="spacer"></div>
    <div id="dummy1" class="dummy"></div>
    <div id="sidebar">
        <br />
        <input type="checkbox"
               style="opacity:0"
               name="noisered"
               value="noisered"
               unchecked>
    </input>
    <br />
    <b>Gene/Transcript: </b></tran> <a class="whatsthis"
    target="_blank"
    href="https://trips.ucc.ie/help/?parent_acc=parent_single_gene&child_acc=child_gene">&#9432;</a>
    <br>
    <input id="tran" name="transcript">
</input>
<br />
<div id="adv_section1">
    <b>Min triplet periodicity score: </b> <a class="whatsthis"
    target="_blank"
    href="https://trips.ucc.ie/help/?parent_acc=parent_single_gene&child_acc=child_max_read_score">&#9432;</a>
    <br>
    <input id="readscore" name="readscore" value="0">
</input>
<br>
<div style="display:none">
    <b>Secondary score: </b> <a class="whatsthis"
    target="_blank"
    href="https://trips.ucc.ie/help/?parent_acc=parent_single_gene&child_acc=child_max_read_score">&#9432;</a>
    <br>
    <input id="secondary_readscore" name="secondary_readscore" value="1">
</input>
<br />
</div>
<br />
<b>Min Read: </b></tran> <a class="whatsthis"
    target="_blank"
    href="https://trips.ucc.ie/help/?parent_acc=parent_single_gene&child_acc=child_single_minread">&#9432;</a>
<br>
<input id="minread" name="minread" value="5">
</input>
<br />
<b>Max Read: </b></tran> <a class="whatsthis"
    target="_blank"
    href="https://trips.ucc.ie/help/?parent_acc=parent_single_gene&child_acc=child_single_maxread">&#9432;</a>
<br>
<input id="maxread" name="maxread" value="150">
</input>
<br />
<b>Highlight sequences list: </b></tran> <a class="whatsthis"
    target="_blank"
    href="https://trips.ucc.ie/help/?parent_acc=parent_single_gene&child_acc=child_seq_element_hili">&#9432;</a>
<br>
<input id="seqhili" name="seqhili" value="">
</input>
<br />
<tr>
    <td>
        <b>Highlight start: </b><a class="whatsthis"
   target="_blank"
   href="https://trips.ucc.ie/help/?parent_acc=parent_comparison&child_acc=child_comp_hili">&#9432;</a>
        <br>
        <input type="text" id="hili_start" name="hili_start" value="0">
    </input>
</tr>
</td>
<br>
<tr>
    <td>
        <b>Highlight stop: </b><a class="whatsthis"
   target="_blank"
   href="https://trips.ucc.ie/help/?parent_acc=parent_comparison&child_acc=child_comp_hili">&#9432;</a>
        <br>
        <input type="text" id="hili_stop" name="hili_stop" value="0">
    </input>
</tr>
</td>
<br />
<b>Offset Direction: </b> <a class="whatsthis"
    target="_blank"
    href="https://trips.ucc.ie/help/?parent_acc=parent_single_gene&child_acc=child_offset_dir">&#9432;</a>
<br>
<input type="radio"
       class="prime_radio fiveprime"
       id="fiveprime"
       value="fiveprime"
       name="primetype"
       checked>
</input>
<label for='fiveprime' class='check_label'>5'</label>
<input type="radio"
       class="prime_radio threeprime"
       id="threeprime"
       value="threeprime"
       name="primetype"
       unchecked>
</input>
<label for='threeprime' class='check_label'>3'</label>
</div>
</div>
<div id="transcripts">
</br>
</br>
</br>
</br>
<input class="line_graph"
       id="line_graph"
       type="checkbox"
       name="namelite"
       value="lite1"
       checked>
</input>
<label for='line_graph' class='checkbox_label'></label>
<label for='line_graph' class='padded_label'>Line Graph</label>
<a class="whatsthis"
   target="_blank"
   href="https://trips.ucc.ie/help/?parent_acc=parent_single_gene&child_acc=child_line">&#9432;</a>
</br>
</br>
<input class="ambig_check"
       id="ambig_check"
       type="checkbox"
       name="ambiguous"
       value="ambiguous"
       unchecked>
</input>
<label for='ambig_check' class='checkbox_label'></label>
<label for='ambig_check' class='padded_label'>Allow ambiguously mapped reads</label>
<a class="whatsthis"
   target="_blank"
   href="https://trips.ucc.ie/help/?parent_acc=parent_single_gene&child_acc=child_ambig">&#9432;</a>
</br>
</br>
<input class="pcr_check"
       id="pcr_check"
       type="checkbox"
       name="pcr"
       value="pcr"
       unchecked>
</input>
<label for='pcr_check' class='checkbox_label'></label>
<label for='pcr_check' class='padded_label'>Allow PCR duplicates</label>
<a class="whatsthis"
   target="_blank"
   href="https://trips.ucc.ie/help/?parent_acc=parent_single_gene&child_acc=child_pcr">&#9432;</a>
</br>
</br>
<div id="adv_section2">
    <input class="cov_check"
           id="cov_check"
           type="checkbox"
           name="coverage"
           value="coverage"
           id="coverageid"
           unchecked>
</input>
<label for='cov_check' class='checkbox_label'></label>
<label for='cov_check' class='padded_label'>Ribo-Seq coverage</label>
<a class="whatsthis"
   target="_blank"
   href="https://trips.ucc.ie/help/?parent_acc=parent_single_gene&child_acc=child_ribocov">&#9432;</a>
</br>
</br>
<input class="nuc_check"
       id="nuc_check"
       type="checkbox"
       name="nucseq"
       value="nucseq"
       id="nucseqid"
       unchecked>
</input>
<label for='nuc_check' class='checkbox_label'></label>
<label for='nuc_check' class='padded_label'>Display nucleotide sequence</label>
<a class="whatsthis"
   target="_blank"
   target="_blank"
   href="https://trips.ucc.ie/help/?parent_acc=parent_single_gene&child_acc=child_nucseq">&#9432;</a>
</br>
</br>
<input class="mismatch_check"
       id="mismatch_check"
       type="checkbox"
       name="mismatchseq"
       value="mismatchseq"
       id="mismatchseqid"
       unchecked>
</input>
<label for='mismatch_check' class='checkbox_label'></label>
<label for='mismatch_check' class='padded_label'>Display mismatches</label>
<a class="whatsthis"
   target="_blank"
   target="_blank"
   href="https://trips.ucc.ie/help/?parent_acc=parent_single_gene&child_acc=child_nucseq">&#9432;</a>
</div>
</div>
<div id = "dialog-2" title = "Choose a transcript..."></div>
<div id = "study_dialog" title = "Study info">
    <table>
    </table>
</div>
<div id="seq_types">
    <center>
        <p>
            <font size="5"><b><u>Seq types</u></b></font>
        </p>
    </center>
    <center>
        <table>
            <!--<input type="button"
       onClick="uncheck_all_riboseq"
       value="Riboseq: Uncheck All Studies    "
       class="button" />-->
            <tr id="1">
                <td>
                    <button class="button all_checks" type="button" onclick='uncheck_all_multi()'>
                        <b>Uncheck All Files With Selected Seq-type</b>
                    </button>
                </td>
                <td>
                    <button class="button all_checks" type="button" onclick='check_all_multi()'>
                        <b>Check All Files with Selected Seq-type</b>
                    </button>
                </td>
            </tr>
        </table>
    </center>
</br>
</div>
<div id="studies">
    <center>
        <p>
            <font size="5"><b><u>Studies</u></b></font>
        </p>
    </center>
</br>
<button class="study_button" type="button" onclick='view_study_info()'>
    <b>View study info</b>
</button>
<center>
    <table>
        <tr id="2">
            <td>
                <button class="button all_checks"
                        type="button"
                        onclick="uncheck_study_multi()">
                    <b>Uncheck All files in Files box</b>
                </button>
            </td>
            <td>
                <button class="button all_checks" type="button" onclick="check_study_multi()">
                    <b>Check All files in Files box</b>
                </button>
            </td>
        </tr>
    </table>
</center>
</br>
</br>
<center>
    <table name="studytable" id="studytable" class="studytable">
        <tr id="tablerow1"></tr>
        <tr id="tablerow2"></tr>
        <tr id="tablerow3"></tr>
        <tr id="tablerow4"></tr>
        <tr id="tablerow5"></tr>
        <tr id="tablerow6"></tr>
        <tr id="tablerow7"></tr>
        <tr id="tablerow8"></tr>
        <tr id="tablerow9"></tr>
        <tr id="tablerow10"></tr>
        <tr id="tablerow11"></tr>
        <tr id="tablerow12"></tr>
        <tr id="tablerow13"></tr>
        <tr id="tablerow14"></tr>
        <tr id="tablerow15"></tr>
        <tr id="tablerow16"></tr>
        <tr id="tablerow17"></tr>
        <tr id="tablerow18"></tr>
    </table>
</center>
</br>
</br>
</div>
<div id="files">
</br>
<p>
    <font size="5" align="center"><b><u>Files</u></b>
    </br>
</font>
</p>
</div>
<div id="buttons"></div>
<div id="view_plot" class="wrapper">
    <button class="button"
            type="button"
            onclick="changebuttonlink(gwips_info)"
            id="query">
        <b>View Plot</b>
    </button>
</div>
{% include 'loading.html' %}
<div class="tab">
    <button id="sc_profile_button"
            class="tablinks"
            onclick="openTab(event, 'sc_profile')">Sub-codon profile</button>
    <button class="tablinks" onclick="openTab(event, 'genome')">Gwips-Viz</button>
    <button class="tablinks" onclick="openTab(event, 'igv')">IGV</button>
</div>
<div id="sc_profile" class="tabcontent">
    <h3>
        Subcodon profile
        <h3>
            <div id="container"></div>
        </div>
        <div id="genome" class="tabcontent">
            <h3>
                Gwips-Viz
                <h3>
                    <iframe id="gwips_iframe"
                            src="https://gwips.ucc.ie"
                            width="1200px"
                            height="660px"></iframe>
                </div>
                <div id="igv" class="tabcontent">
                    <h3>
                        IGV
                        <h3>
                            <div id="myDiv"
                                 style="padding-top: 50px;
                                        padding-bottom: 20px;
                                        height: auto"></div>
                        </div>
                        <script type="text/javascript"
                                src="{{ url_for('static', filename='js/common.js') }}"></script>
                        <script type="text/javascript">
// When the page is completely loaded call the update_colours function
$(document).ready(function(){
	update_colours()
})

var alt_file_list = [];

var user_transcript = "{{ user_transcript|safe }}";
var user_readscore = "{{ user_readscore|safe }}";
var user_minread = "{{ user_minread|safe }}";
var user_maxread = "{{ user_maxread|safe }}";
var user_hili_starts = {{ user_hili_starts|safe }};
var user_hili_stops = {{ user_hili_stops|safe }};
var user_dir = "{{ user_dir|safe }}";
var user_line_graph = "{{ user_line_graph|safe }}";
var user_ambig = "{{ user_ambig|safe }}";
var user_cov = "{{ user_cov|safe }}";
var user_nuc = "{{ user_nuc|safe }}";
var user_short = "{{ user_short|safe }}";
var user_files = {{ user_files|safe }};
var user_ribo_studies = {{ user_ribo_studies|safe }};
var user_proteomics_studies = {{ user_proteomics_studies|safe }};
var user_rna_studies = {{ user_rna_studies|safe }};
var user_crd = "{{ user_crd|safe }}";
var defaulttran = "{{ default_tran|safe }}";


rownum = 1
rowcount = -1



var seq_types = {{  seq_types|safe  }};
for (var i in seq_types) {
	//console.log("SEQ TYPES "+seq_types[i])
	//if the seq type is riboseq make it checked by default
	if (seq_types[i] == "riboseq") {
		checked = "checked"
	}
	else {
		checked = "unchecked"
	};

	// if seq type is riboseq or rnaseq set a proper display name, else just use seq_type
	if (seq_types[i] == "riboseq") {
		display_name = "Ribo-Seq"
	}
	else if (seq_types[i] == "rnaseq") {
		display_name = "mRNA-Seq"
	}
	else {
		display_name = seq_types[i]
	}

	var radio_seq_types = "<input "+checked+" type='radio' class='seq_type_radio all_checks' id='"+seq_types[i]+"' name='seq_type' value="+seq_types[i]+" /></b>   </input><label for='"+seq_types[i]+"' id='label"+seq_types[i]+"' class='radio_label'>"+display_name+"</label>";
	var seq_types_div = document.getElementById('seq_types');
	seq_types_div.innerHTML += radio_seq_types;
}



var studies_dict = {{  studies_dict|safe  }};
var studyinfo_dict = {{  studyinfo_dict|safe  }};
for (var key in studies_dict)  {
		rowcount += 1;
		if (rowcount == 8) {
				rowcount = 0
				rownum += 1
				};
		var radio_study = "<input type='radio' checked class='study_radio all_checks' id='study"+key+"' name='content_type' value=study"+key+" /></b></input><label for='study"+key+"' id='label"+key+"' class='radio_label'>"+studies_dict[key]["study_name"]+"</label>";
		//console.log("LABEL ID IS label"+key);
		addRow(rownum, radio_study)
		for (i in seq_types) {
			var seq_type = seq_types[i]
			var files_div = document.getElementById('files');
			var new_show_div = document.createElement(studies_dict[key]["study_name"]);
			//console.log("Seq type is "+seq_type)
			new_show_div.className = 'show';
			new_show_div.id = seq_type+"study"+key;
			tablehtml = "<table class='file_table' id="+key+seq_type+" align='center'></table>";
			new_show_div.innerHTML = tablehtml;
			files_div.appendChild(new_show_div);
		}
		};

// accepted files structure, key is file_type,values are file_id's (which are themselves keys) which point to a dictionary with the following: {"study_id":row[1],"file_name":row[2],"file_description":row[3],"file_type":row[4]}
var accepted_files = {{ accepted_files|safe }};


// keys in this are seq types with values being a list of study id's that have at least one file of that seq type
// this is used to hide studies that don't have any files of the selected seq type.
var seq_study_dict = {};


// populate the ribostudies and rnastudies div with file names and descriptions.
for (var filetype in accepted_files) {
		//console.log("filetype ", filetype);
		for (var study_id in accepted_files[filetype]) {
				//console.log("Study id: ",study_id);
				i=0
				var file_names = [];
				for (var file_id in accepted_files[filetype][study_id]) {
					file_name = accepted_files[filetype][study_id][file_id]["file_name"]
					file_names.push([file_name, file_id])
					}
				console.log("File names", file_names);
				file_names = file_names.sort(function(a, b) {
					return a[0].localeCompare(b[0]);
				})
				
				var sorted_file_list = [];
				for (var z in file_names) {
					var tup = file_names[z];
					sorted_file_list.push(tup[1]);
					}
				console.log("sorted file_list", sorted_file_list);
				for (var file_id in sorted_file_list) {
						file_id = sorted_file_list[file_id]
						file_name = accepted_files[filetype][study_id][file_id]["file_name"];
						file_description = accepted_files[filetype][study_id][file_id]["file_description"];
						if (!(filetype in seq_study_dict)) {
							seq_study_dict[filetype] = [];
						};
						//console.log("FILE ID LIST IS "+accepted_files[filetype][study_id])
						if (accepted_files[filetype][study_id].length != 0) {
							if (!(study_id in seq_study_dict[filetype])) {
								//console.log("PUSHING STUDY ID"+study_id);
								seq_study_dict[filetype].push(study_id);
								//console.log("Seq study dict now "+seq_study_dict[filetype])
						}
					};


						var table = document.getElementById(study_id+filetype);
						var tr = document.createElement("tr");
						var td = document.createElement("td");
						// riboseq is checked by default all others not
						if (filetype == "riboseq") {
								addItOnce(alt_file_list, file_id);
								td.innerHTML = ("<td><input id='"+file_name+file_id+"_check' type='checkbox' class='file_check file_"+file_id+" all_checks "+filetype+"_file "+filetype+"_study"+study_id+"'  name="+file_id+" value=inputfile checked></input>    <label for='"+file_name+file_id+"_check' class='filetable_label'></label> <label for='"+file_name+file_id+"_check' class='padded_label'><b>"+file_name+"</b>:"+file_description+"</label></td>");
								}
						else {
								td.innerHTML = ("<td><input id='"+file_name+file_id+"_check' type='checkbox' class='file_check file_"+file_id+" all_checks "+filetype+"_file "+filetype+"_study"+study_id+"'  name="+file_id+" value=inputfile unchecked></input>    <label for='"+file_name+file_id+"_check' class='filetable_label'></label> <label for='"+file_name+file_id+"_check' class='padded_label'><b>"+file_name+"</b>:"+file_description+"</label></td>");
								}
						//tr.appendChild();

						if (i != 2){
								table.appendChild(td);
								i+= 1;
												}
						else if (i == 2){
								table.appendChild(td);
								table.appendChild(tr);
								i = 0
								}

																						}
														}
								};




// If the user presses the enter key call the #query function (generates the plot)
$(document).keypress(function(e){
		if (e.which == 13){
				remove_user_short();
				}
		});

		
		
// checks all checkboxes with the <seq_type>_study<study_id> class, the study_id is gotten from the value of the currently selected study
function check_study_multi(){
		var curstudy = $('input[name="content_type"]:checked').val();
		var seq_type = document.querySelector('input[name="seq_type"]:checked').value;
		$('input.'+seq_type+'_'+curstudy+'').prop('checked',true);
		update_colours();
		$('input.'+seq_type+'_'+curstudy+'').each(function(){
			var file_id = $(this).attr('name');
			//alt_file_list.push(file_id);
			addItOnce(alt_file_list, file_id);
			//console.log("Adding file id "+file_id);
		});
};

function uncheck_study_multi(){
		var curstudy = $('input[name="content_type"]:checked').val();
		var seq_type = document.querySelector('input[name="seq_type"]:checked').value;
		$('input.'+seq_type+'_'+curstudy+'').prop('checked',false);
		update_colours();
		
		$('input.'+seq_type+'_'+curstudy+'').each(function(){
			var file_id = $(this).attr('name');
			removeItemOnce(alt_file_list, file_id);
			
		});
};		

// Unchecks every checkbox that has the class <seq_type>_file
function uncheck_all_multi(){
		var seq_type = document.querySelector('input[name="seq_type"]:checked').value;
		$('input.'+seq_type+'_file').prop('checked',false);
		update_colours();
		$('input.'+seq_type+'_file').each(function(){
			var file_id = $(this).attr('name');
			removeItemOnce(alt_file_list, file_id);
			
		});
};

// Checks every checkbox that has the class <seq_type>_file
function check_all_multi(){
		var seq_type = document.querySelector('input[name="seq_type"]:checked').value;
		$('input.'+seq_type+'_file').prop('checked',true);
		update_colours();
		$('input.'+seq_type+'_file').each(function(){
			var file_id = $(this).attr('name');
			addItOnce(alt_file_list, file_id);
			//console.log("Adding file id "+file_id);
		});
};





		
// This is called by pressing enter or view plot, it's function is to remove any short code passed by the user as a html argument. The reason for this is
// if someone clicks on a short link there willl be a 'short_code=xxxx' in the html_arguments. If they then change a value and view the plot again the script
// will see the short code in the html arguments and assume this short_code applies to the new graph and so the short code will not change. By routing the
// view plot and enter buttons through this function first we can strip the html_args of the short_code argument and thus force the script to create a new one.
function remove_user_short() {
		changebuttonlink(gwips_info)
		user_short = "None";
		//alert("The function called 'function_one' has been called.");
		$("#query").click();
		}

var url = "/query"


// hide advanced sections from user if they have not turned on the advanced setting
var advanced = "{{ advanced|safe }}";
if (advanced == "False") {
		var adv_section1 = document.getElementById('adv_section1');
		adv_section1.style.display = 'none';
		var adv_section2 = document.getElementById('adv_section2');
		adv_section2.style.display = 'none';
		}


// This is called when the user passes a gene, shows a popup with multiple transcripts
$("#loading-div-background").css({ opacity: 0.7 });
$(function() {
		$( "#dialog-2" ).dialog({
				autoOpen: false,
				modal: true,
				minWidth: 800,
				buttons: {
						Submit: function() {//var transcript = document.getElementById("transcript_sel").value();
																var transcript = document.querySelector('input[name=transcript_sel]:checked').value;
																document.getElementById('tran').value = transcript;
																$(this).dialog("close");
																$("#query").click();
																}
						},
				title: "Choose a transcript. . .",
				position: {
						my: "center",
						at: "center"
								}
				});
				$( "#query4" ).click(function() {
						$( "#dialog-2" ).dialog("open");
						});
				});


// Query function, gathers information on settings configured by the user and passes it to python for a plot to be generated
$("#query").click(function() {
	console.log("Query Called");
	$("#loading-div-background").show();
	$("#container").hide();
	var nanobar = new Nanobar({bg: '#77abff'});
	var primetype = document.querySelector('input[name=primetype]:checked').value;
	var transcript = $('input:text[name=transcript]').val();
	
	var varlite = $('input:checkbox[name=namelite]:checked').val();
	var minread = $('input:text[name=minread]').val();
	var maxread = $('input:text[name=maxread]').val();
	var seqhili = $('input:text[name=seqhili]').val();
	var minfiles = 0    //$('input:text[name=minfiles]').val();
	var ribocoverage = $('input:checkbox[name=coverage]:checked').val();
	var nucseq = $('input:checkbox[name=nucseq]:checked').val();
	var mismatches = $('input:checkbox[name=mismatchseq]:checked').val();
	var preprocess = $('input:checkbox[name=preprocess]:checked').val();
	var color_readlen_dist = $('input:checkbox[name=readlen_dist]:checked').val();
	var noisered = $('input:checkbox[name=noisered]:checked').val();
	var ambiguous = $('input:checkbox[name=ambiguous]:checked').val();
	var pcr = $('input:checkbox[name=pcr]:checked').val();
	var readscore = $('input:text[name=readscore]').val();
	var secondary_readscore = $('input:text[name=secondary_readscore]').val();
	var organism = "{{  organism|safe  }}";
	var transcriptome = "{{  transcriptome|safe  }}";
	var hili_start = $('input:text[name=hili_start]').val();
	var hili_stop = $('input:text[name=hili_stop]').val();
	
	gwips_clade = gwips_info["database"];
	gwips_org = gwips_info["organism"];
	gwips_db = gwips_info["clade"];
	iframe_link = "https://gwips.ucc.ie/cgi-bin/hgTracks?hideTracks=1&db="+gwips_clade+"&org="+gwips_org+"&db="+gwips_db+"&wgEncodeGencodeBasicV25=pack&Global_mRNACov=pack&Global_RiboPro=pack&g=wgEncodeGencodeBasicV25&i=" + transcript + "";
	console.log("setting gwips iframe to link "+iframe_link);
	document.getElementById('gwips_iframe').src = iframe_link;
	
	
	
	
	var qu = {"minfiles":minfiles,"transcript":transcript, "varlite": varlite, "minread":minread, "maxread":maxread,"noisered":noisered,  "ribocoverage": ribocoverage,
			"ambiguous": ambiguous,"pcr":pcr,"organism":organism,"transcriptome":transcriptome,"readscore":readscore, "primetype":primetype,"preprocess":preprocess,
			"nucseq":nucseq, "user_hili_starts":user_hili_starts, "user_hili_stops":user_hili_stops,"color_readlen_dist":color_readlen_dist,
			"user_short":user_short,"advanced":advanced,"seqhili":seqhili, "secondary_readscore":secondary_readscore,"mismatches":mismatches,
			"hili_start":hili_start,"hili_stop":hili_stop};
	console.log("Building file_list");
	//var file_list = [];
	//for (var filetype in accepted_files) {
	//		for (var study_id in accepted_files[filetype]) {
	//				for (var file_id in accepted_files[filetype][study_id]) {
	//						var newelement=$('input:checkbox[name='+file_id+']:checked').val();
	//						if (newelement == 'inputfile'){
	//								file_list.push(file_id);
	//																					};
	//																																 }
	//																										}
	//																			};
	//qu["file_list"] = file_list;
	qu["file_list"] = alt_file_list;
	//console.log("len of File list is "+file_list.length);
	//console.log("len of Alt file list is "+alt_file_list);
	if (alt_file_list.length == 1) {
		loading_header.innerText = "Loading data from "+alt_file_list.length+" file";}
	else {
		loading_header.innerText = "Loading data from "+alt_file_list.length+" files";}
	$.ajax({
		type: "POST",
		async:true,
		contentType: "application/json; charset=utf-8",
		url: url,
		data: JSON.stringify(qu),
		dataType: "html",
		success: function(data) {
					
					
						if (data.slice(0, 11) == "TRANSCRIPTS"){
							initBrowser(transcript);
							var splitdata = data.split(":");
							populate_jquery_dialog(splitdata);
							$("#myTable").tablesorter({sortList: [[4,0]]});
							$( "#dialog-2" ).dialog( "open" );
							}
			
						else if (data.slice(0, 17) == "QUANT_TRANSCRIPTS"){
							var splitdata = data.split(":");
							populate_jquery_dialog_quant(splitdata);
							$("#myTable").tablesorter({sortList: [[6,0]]});
							$( "#dialog-2" ).dialog( "open" );
							}

						else{
							var graph = $("#container");
							graph.html(data);
							
							console.log("clicking sc_profile");
							document.getElementById("sc_profile_button").click();
							$("#loading-div-background").hide();
							$("#container").show();
							$('html, body').animate({scrollTop: $("#container").offset().bottom}, 200);
							}
						document.getElementById('container').scrollIntoView();
					
					

		}
	 });
});




function update_progress(status_url, nanobar) {
	// send GET request to status URL
	$.getJSON(status_url, function(data) {
		// update UI
		percent = parseInt(data['current'] * 100 / data['total']);
		nanobar.go(percent);
		//$(status_div.childNodes[1]).text(percent + '%');
		//$(status_div.childNodes[2]).text(data['status']);
		$("#loading_header").text(data['status'])
		if (data['state'] != 'PENDING' && data['state'] != 'PROGRESS') {
			var graph = $("#container");
			graph.html(data['result']);
			$("#loading-div-background").hide();
			$("#container").show();
			$('html, body').animate({scrollTop: $("#container").offset().bottom}, 200);
			document.getElementById('container').scrollIntoView();
			}
		else {
			// rerun in 2 seconds
			setTimeout(function() {
				update_progress(status_url, nanobar);
				}, 500);
			}
		});
	}


// if the user passes a list of files, or studies uncheck everything then recheck only those checkboxes which the user passed
if (user_files.length != 0 || user_ribo_studies.length != 0 || user_proteomics_studies.length != 0 || user_rna_studies.length != 0) {
		for (i in seq_types) {
			var seq_type = seq_types[i];
			$('input.'+seq_type+'_file').prop('checked',false);
			console.log("User files submitted setting alt file list to empty");
			alt_file_list = [];
		}

		if (user_files.length != 0) {
				console.log("Adding individual user files to alt file list");
				for (i in user_files) {
						$('input.file_'+user_files[i]+'').prop('checked',true);
						addItOnce(alt_file_list, user_files[i]);
						}};

		if (user_ribo_studies.length != 0) {
				for (i in user_ribo_studies) {
						console.log("Adding riboseq study to alt file list", user_ribo_studies[i]);
						//console.log("USER RIBO STUDIES "+user_ribo_studies[i]);
						$('input.riboseq_study'+user_ribo_studies[i]+'').prop('checked',true);
						$('input.riboseq_study'+user_ribo_studies[i]+'').each(function(){
							var file_id = $(this).attr('name');
							addItOnce(alt_file_list, file_id);
						});
					}};
						
		if (user_proteomics_studies.length != 0) {
			for (i in user_proteomics_studies) {
						console.log("Adding proteomics study to alt file list", user_proteomics_studies[i]);
						//console.log("USER RIBO STUDIES "+user_ribo_studies[i]);
						$('input.proteomics_study'+user_proteomics_studies[i]+'').prop('checked',true);
						$('input.proteomics_study'+user_proteomics_studies[i]+'').each(function(){
							var file_id = $(this).attr('name');
							addItOnce(alt_file_list, file_id);
						});
					}};

		if (user_rna_studies.length != 0) {
				for (i in user_rna_studies) {
						console.log("Adding rna study to alt file list", user_rna_studies[i]);
						$('input.rnaseq_study'+user_rna_studies[i]+'').prop('checked',true);
						$('input.rnaseq_study'+user_rna_studies[i]+'').each(function(){
							var file_id = $(this).attr('name');
							addItOnce(alt_file_list, file_id);
						});
					}};
		}


if (user_readscore != "None") {
		document.getElementById('readscore').value = user_readscore;
		}

if (user_minread != "None") {
		document.getElementById('minread').value = user_minread;
		}

if (user_maxread != "None") {
		document.getElementById('maxread').value = user_maxread;
		}

if (user_line_graph == "F") {
		$('.line_graph').attr('checked', false);
		}

if (user_ambig == "T") {
		$('.ambig_check').attr('checked', true);
		}
if (user_cov == "T") {
		$('.cov_check').attr('checked', true);
		}

if (user_nuc == "T") {
		$('.nuc_check').attr('checked', true);
		}

if (user_crd == "T") {
		//console.log("Setting crd to true");
		$('.crd').attr('checked',true);
		}



if (user_dir != "None") {
		if (user_dir == "fiveprime") {
				$('.fiveprime').attr('checked', true);
												 }
		else if (user_dir == "threeprime") {
				$('.threeprime').attr('checked', true);
															}
												};

if (user_transcript != "None") {
		document.getElementById('tran').value = user_transcript;
		// If a transcript is passed by the user then automatically produce a graph
		$("#query").click();
		}
else {
		document.getElementById('tran').value = defaulttran;
		}


//Update the options box on page load and load files in file box
$(document).ready(function(){
	var n = document.querySelector('input[name="content_type"]:checked').value;
	var seq_type = document.querySelector('input[name="seq_type"]:checked').value;
	for (var study_id in studies_dict) {
		$("#label"+study_id+"").css({ opacity: 0.2 });
	}
	for (file_type in seq_study_dict) {
		if (file_type == seq_type) {
			//console.log("File type is seq_type "+file_type+"  "+seq_type);
			for (var sub_study_id in seq_study_dict[file_type]) {
				$("#label"+seq_study_dict[file_type][sub_study_id]+"").css({ opacity: 1 });
			}
		}
	}

	var selector = "#"+seq_type+ n; //Create a selector based on the id
	$(".show").hide(); //Hide all the checkbox containers
	$(selector).show(); //Show only the related container
})


// This section listens for clicks from any radio/checkbox with the all_checks class, if it detects a change it goes through
// all studies to see which if any of them have riboseq/rnaseq files checked, if so it changes the border colour
$('.all_checks').change(function() {
		for (var key in studies_dict)  {
				if (($('.rnaseq_study'+key+':checked').length == 0) && ($('.riboseq_study'+key+':checked').length == 0)) {
						document.getElementById('label'+key+'').style.borderColor='#e2e2e2';
						}
				else {
						document.getElementById('label'+key+'').style.borderColor='#77abff';
						}
		}
		});


$('#files').on('change', '.file_check', function() {
	//console.log("check function called")
	var file_id = $(this).attr('name');
	if (this.checked) {
		//console.log("adding file id "+ $(this).attr('name'))
		addItOnce(alt_file_list, file_id);
		}
	else {
		removeItemOnce(alt_file_list, file_id);
		};
});
		
function removeItemOnce(arr, value) {
	var index = arr.indexOf(value);
	if (index > -1) {
		//console.log("Poppin file id "+value);
		arr.splice(index, 1);
	}
	return arr;
};
	
	
	
function addItOnce(arr, value) {
	var index = arr.indexOf(value);
	if (index < 0) {
		//console.log("adding file id "+value);
		arr.push(value);
	}
	return arr;
};		

// This function is called whenever the check/uncheck all buttons are pressed, when called it goes through
// all studies to see which if any of them have riboseq/rnaseq files checked, if so it changes the border colour
function update_colours() {
		for (var key in studies_dict)  {
				if (($('.rnaseq_study'+key+':checked').length == 0) && ($('.riboseq_study'+key+':checked').length == 0)) {
						document.getElementById('label'+key+'').style.borderColor='#e2e2e2';
						}
				else {
						document.getElementById('label'+key+'').style.borderColor='#77abff';
						}
				}
				}

				
function openTab(evt, tabName) {
	var i, tabcontent, tablinks;
	
	tabcontent = document.getElementsByClassName("tabcontent");
	for (i = 0; i < tabcontent.length; i++) {
		tabcontent[i].style.display = "none";
	}
	
	tablinks = document.getElementsByClassName("tablinks");
	for (i = 0; i < tablinks.length; i++) {
		tablinks[i].className = tablinks[i].className.replace(" active","");
	}
	
	document.getElementById(tabName).style.display = "block";
	evt.currentTarget.className += 'active';
}
				
				

function initBrowser(gene) {
	igv.removeAllBrowsers()
	var div,options,browser;
	div = document.getElementById("myDiv");
	options = {
		genome: "hg38",
		locus: gene,
		tracks: [
					{
						url: "{{ url_for('static', filename='global_ribo.bw') }}",
						name: 'Global Ribo-Seq'
					},
					{
						url: "{{ url_for('static', filename='global_rna.bw') }}",
						name: 'Global mRNA'
					},
					{
						"name": "Gencode v40",
      "type": "annotation",
      "format": "gff3",
      "displayMode": "EXPANDED",
      "height": 300,
      "url": "https://s3.amazonaws.com/igv.org.genomes/hg38/gencode.v40.annotation.gff3.gz",
      "indexURL": "https://s3.amazonaws.com/igv.org.genomes/hg38/gencode.v40.annotation.gff3.gz.tbi",
      "visibilityWindow": 1000000
					}
		]
	};
	browser = igv.createBrowser(div, options);
}

document.addEventListener("DOMContentLoaded", function () {
	initBrowser("B2M");
});





                        </script>
                    {% endblock %}
